File: src/service/AppService.js
const Docker = require('dockerode');
const pm2 = require('pm2');
const App = require('../model/App');

// Connect ke Docker daemon via Unix socket
const docker = new Docker({ socketPath: '/var/run/docker.sock' });

class AppService {
  /**
   * Connect ke PM2 daemon
   */
  static connectPM2() {
    return new Promise((resolve, reject) => {
      pm2.connect((err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  /**
   * Disconnect dari PM2 daemon
   */
  static disconnectPM2() {
    try {
      pm2.disconnect();
    } catch (e) {
      // Ignore
    }
  }

  /**
   * Ambil daftar semua apps (Docker containers + PM2 processes)
   */
  static async getApps(user) {
    const apps = [];

    // 1. Ambil Docker containers
    try {
      const containers = await docker.listContainers({ all: true });
      const dockerApps = containers.map(c => {
        const name = (c.Names?.[0] || '').replace(/^\//, '');
        const state = c.State;

        let status = 'unknown';
        if (state === 'running') status = 'online';
        else if (state === 'exited') status = 'stopped';
        else if (state === 'paused') status = 'stopped';
        else if (state === 'restarting') status = 'processing';
        else status = state;

        return {
          name,
          status,
          cpu: 0,
          memory: 0,
          uptime: c.Created ? c.Created * 1000 : null,
          pm_id: c.Id?.substring(0, 12) || '',
          cwd: '',
          containerId: c.Id,
          image: c.Image,
          type: 'docker',
          ports: c.Ports || [],
        };
      });
      apps.push(...dockerApps);
    } catch (err) {
      console.error('Gagal ambil Docker containers:', err.message);
    }

    // 2. Ambil PM2 processes
    try {
      await this.connectPM2();
      const list = await new Promise((resolve, reject) => {
        pm2.list((err, list) => {
          if (err) reject(err);
          else resolve(list);
        });
      });

      const pm2Apps = list.map(app => {
        const name = app.name;
        // Skip jika sudah ada di Docker (hindari duplikasi)
        if (apps.find(a => a.name === name)) return null;

        return {
          name,
          status: app.pm2_env?.status || 'unknown',
          cpu: app.monit?.cpu || 0,
          memory: app.monit?.memory || 0,
          uptime: app.pm2_env?.pm_uptime,
          pm_id: app.pm2_env?.pm_id || 0,
          cwd: app.pm2_env?.pm_cwd || '',
          type: 'pm2',
        };
      }).filter(Boolean);

      apps.push(...pm2Apps);
    } catch (err) {
      console.error('Gagal ambil PM2 processes:', err.message);
    } finally {
      this.disconnectPM2();
    }

    // Sync apps ke database
    await App.syncFromPM2(apps);

    // Filter by allowed apps jika bukan superadmin
    if (user.role !== 'superadmin' && user.allowedApps) {
      return apps.filter(app => user.allowedApps.includes(app.name));
    }

    return apps;
  }

  /**
   * Lakukan aksi start/stop/restart pada app (Docker atau PM2)
   */
  static async doAction(appName, action, user) {
    // Check permission untuk non-superadmin
    if (user.role !== 'superadmin' && user.allowedApps) {
      if (!user.allowedApps.includes(appName)) {
        throw new Error('Anda tidak memiliki akses ke aplikasi ini');
      }
    }

    // Cari app di Docker dulu
    const containers = await docker.listContainers({ all: true });
    const container = containers.find(c => {
      const name = (c.Names?.[0] || '').replace(/^\//, '');
      return name === appName;
    });

    if (container) {
      // Handle Docker container
      const cont = docker.getContainer(container.Id);

      switch (action) {
        case 'start':
          await cont.start();
          break;
        case 'stop':
          await cont.stop();
          break;
        case 'restart':
          await cont.restart();
          break;
        default:
          throw new Error(`Aksi '${action}' tidak didukung. Gunakan start, stop, atau restart.`);
      }

      return { action, container: appName, type: 'docker' };
    }

    // Jika tidak ada di Docker, coba PM2
    await this.connectPM2();
    try {
      return await new Promise((resolve, reject) => {
        pm2[action](appName, (err, result) => {
          if (err) reject(err);
          else resolve({ action, container: appName, type: 'pm2' });
        });
      });
    } finally {
      this.disconnectPM2();
    }
  }

  /**
   * Ambil stats real-time (CPU & Memory) untuk satu container
   */
  static async getContainerStats(containerName) {
    const containers = await docker.listContainers({ all: false });
    const container = containers.find(c => {
      const name = (c.Names?.[0] || '').replace(/^\//, '');
      return name === containerName;
    });

    if (!container) return null;

    const cont = docker.getContainer(container.Id);
    const stats = await cont.stats({ stream: false });

    // Hitung CPU usage
    const cpuDelta = stats.cpu_stats.cpu_usage.total_usage - stats.precpu_stats.cpu_usage.total_usage;
    const systemDelta = stats.cpu_stats.system_cpu_usage - stats.precpu_stats.system_cpu_usage;
    const numCpus = stats.cpu_stats.online_cpus || 1;
    const cpuPercent = systemDelta > 0 ? (cpuDelta / systemDelta) * numCpus * 100 : 0;

    // Hitung Memory usage
    const memUsed = stats.memory_stats.usage - (stats.memory_stats.stats?.cache || 0);
    const memLimit = stats.memory_stats.limit;

    return {
      cpu: Math.round(cpuPercent * 100) / 100,
      memory: memUsed,
      memoryPercent: memLimit > 0 ? Math.round((memUsed / memLimit) * 10000) / 100 : 0,
    };
  }
}

module.exports = AppService;

File: internal/service/app_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"log"
	"strings"

	"github.com/tragasolusi/pm2-manager-api/internal/auth"
	"github.com/tragasolusi/pm2-manager-api/internal/docker"
	"github.com/tragasolusi/pm2-manager-api/internal/model"
	"github.com/tragasolusi/pm2-manager-api/internal/pm2"
	"github.com/tragasolusi/pm2-manager-api/internal/repository"
)

var (
	ErrAppNotFound    = errors.New("aplikasi tidak ditemukan")
	ErrActionInvalid  = errors.New("action harus start, stop, atau restart")
	ErrNoAccessToApp  = errors.New("anda tidak memiliki akses ke aplikasi ini")
	ErrAppUnavailable = errors.New("aplikasi tidak tersedia di PM2 atau Docker")
)

// AppView is the public-facing summary of an app.
type AppView struct {
	Name        string  `json:"name"`
	Status      string  `json:"status"`
	CPU         float64 `json:"cpu"`
	Memory      float64 `json:"memory"`
	Uptime      any     `json:"uptime"`      // ms (docker) or epoch (pm2)
	CWD         string  `json:"cwd"`
	Type        string  `json:"type"`         // "docker" | "pm2"
	ContainerID string  `json:"containerId,omitempty"`
	Image       string  `json:"image,omitempty"`
	PmID        any     `json:"pm_id"`
}

type AppService struct {
	dockerCli *docker.Client
	pm2Cli    *pm2.Client
	apps      *repository.AppRepository
}

func NewAppService(d *docker.Client, p *pm2.Client, a *repository.AppRepository) *AppService {
	return &AppService{dockerCli: d, pm2Cli: p, apps: a}
}

// GetApps returns the full list (Docker + PM2), filtered by allowedApps for non-admin.
func (s *AppService) GetApps(ctx context.Context, claims *auth.Claims) ([]AppView, error) {
	views := make([]AppView, 0)
	seen := map[string]bool{}

	// 1. Docker containers
	if s.dockerCli != nil {
		cs, err := s.dockerCli.List(ctx)
		if err != nil {
			log.Printf("docker list error: %v", err)
		} else {
			for _, c := range cs {
				views = append(views, AppView{
					Name:        c.Name,
					Status:      mapDockerStatus(c.State),
					CWD:         "",
					Type:        "docker",
					ContainerID: c.ID,
					Image:       c.Image,
					Uptime:      c.Created * 1000,
					PmID:        shortID(c.ID),
				})
				seen[c.Name] = true
			}
		}
	}

	// 2. PM2 processes
	if s.pm2Cli != nil {
		procs, err := s.pm2Cli.List(ctx)
		if err != nil {
			log.Printf("pm2 list error: %v", err)
		} else {
			for _, p := range procs {
				if seen[p.Name] {
					continue
				}
				views = append(views, AppView{
					Name:   p.Name,
					Status: p.Pm2Env.Status,
					CPU:    p.Monit.CPU,
					Memory: p.Monit.Memory,
					Uptime: p.Pm2Env.PmUptime,
					CWD:    p.Pm2Env.PmCwd,
					Type:   "pm2",
					PmID:   p.PmID,
				})
			}
		}
	}

	// Sync to DB
	if err := s.syncToDB(ctx, views); err != nil {
		log.Printf("sync apps to db: %v", err)
	}

	// Filter for non-admin
	if claims != nil && claims.Role != "superadmin" && len(claims.AllowedApps) > 0 {
		allowed := make(map[string]bool, len(claims.AllowedApps))
		for _, a := range claims.AllowedApps {
			allowed[a] = true
		}
		filtered := views[:0]
		for _, v := range views {
			if allowed[v.Name] {
				filtered = append(filtered, v)
			}
		}
		views = filtered
	}

	return views, nil
}

func (s *AppService) syncToDB(ctx context.Context, views []AppView) error {
	models := make([]model.App, 0, len(views))
	for _, v := range views {
		pmID := ""
		switch id := v.PmID.(type) {
		case string:
			pmID = id
		case int:
			pmID = fmt.Sprintf("%d", id)
		case float64:
			pmID = fmt.Sprintf("%.0f", id)
		}
		models = append(models, model.App{
			Name: v.Name,
			CWD:  v.CWD,
			PMID: pmID,
		})
	}
	return s.apps.SyncFromPM2(ctx, models)
}

// DoAction runs start/stop/restart on the named app, preferring Docker.
func (s *AppService) DoAction(ctx context.Context, name, action string, claims *auth.Claims) error {
	if !isValidAction(action) {
		return ErrActionInvalid
	}
	if claims != nil && claims.Role != "superadmin" && !contains(claims.AllowedApps, name) {
		return ErrNoAccessToApp
	}

	// Try Docker first
	if s.dockerCli != nil {
		ctr, err := s.dockerCli.Get(ctx, name)
		if err != nil {
			log.Printf("docker get %s: %v", name, err)
		} else if ctr != nil {
			return s.dockerCli.Action(ctx, ctr.ID, action)
		}
	}

	// Fallback to PM2
	if s.pm2Cli != nil {
		if err := s.pm2Cli.Action(ctx, action, name); err == nil {
			return nil
		} else {
			return fmt.Errorf("gagal %s '%s': %w", action, name, err)
		}
	}
	return ErrAppUnavailable
}

func isValidAction(a string) bool {
	return a == "start" || a == "stop" || a == "restart"
}

func contains(haystack []string, needle string) bool {
	for _, h := range haystack {
		if strings.EqualFold(h, needle) {
			return true
		}
	}
	return false
}

func mapDockerStatus(state string) string {
	switch state {
	case "running":
		return "online"
	case "exited", "paused", "dead":
		return "stopped"
	case "restarting":
		return "processing"
	default:
		return state
	}
}

func shortID(id string) string {
	if len(id) >= 12 {
		return id[:12]
	}
	return id
}
